File: app/src/main/java/com/krystianwsul/checkme/gui/projects/ShowProjectActivity.java
package com.krystianwsul.checkme.gui.projects;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.design.widget.FloatingActionButton;
import android.support.design.widget.TextInputLayout;
import android.support.v4.app.LoaderManager;
import android.support.v4.content.Loader;
import android.support.v7.app.ActionBar;
import android.support.v7.widget.Toolbar;
import android.text.Editable;
import android.text.TextUtils;
import android.text.TextWatcher;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.EditText;

import com.krystianwsul.checkme.R;
import com.krystianwsul.checkme.gui.AbstractActivity;
import com.krystianwsul.checkme.gui.DiscardDialogFragment;
import com.krystianwsul.checkme.gui.friends.UserListFragment;
import com.krystianwsul.checkme.loaders.ShowProjectLoader;

import junit.framework.Assert;

public class ShowProjectActivity extends AbstractActivity implements LoaderManager.LoaderCallbacks<ShowProjectLoader.Data> {
    private static final String PROJECT_ID_KEY = "projectId";

    private static final String DISCARD_TAG = "discard";

    @Nullable
    private String mProjectId;

    private TextInputLayout mToolbarLayout;
    private EditText mToolbarEditText;

    private ShowProjectLoader.Data mData;

    private UserListFragment mUserListFragment;

    @Nullable
    private Bundle mSavedInstanceState;

    private final DiscardDialogFragment.DiscardDialogListener mDiscardDialogListener = this::finish;

    public static Intent newIntent(@NonNull Context context, @NonNull String projectId) {
        Assert.assertTrue(!TextUtils.isEmpty(projectId));

        Intent intent = new Intent(context, ShowProjectActivity.class);
        intent.putExtra(PROJECT_ID_KEY, projectId);
        return intent;
    }

    public static Intent newIntent(@NonNull Context context) {
        return new Intent(context, ShowProjectActivity.class);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.menu_save, menu);
        return true;
    }

    @Override
    public boolean onPrepareOptionsMenu(Menu menu) {
        menu.findItem(R.id.action_save).setVisible(mData != null);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.action_save: {
                Assert.assertTrue(mData != null);

                if (updateError())
                    break;

                getSupportLoaderManager().destroyLoader(0);

                mUserListFragment.save(mToolbarEditText.getText().toString());

                finish();

                break;
            }
            case android.R.id.home: {
                if (tryClose())
                    finish();

                break;
            }
            default:
                throw new UnsupportedOperationException();
        }
        return true;
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_show_project);

        mSavedInstanceState = savedInstanceState;

        Toolbar toolbar = (Toolbar) findViewById(R.id.toolbar);
        Assert.assertTrue(toolbar != null);

        setSupportActionBar(toolbar);

        ActionBar actionBar = getSupportActionBar();
        Assert.assertTrue(actionBar != null);

        actionBar.setDisplayHomeAsUpEnabled(true);
        actionBar.setHomeAsUpIndicator(R.drawable.ic_close_white_24dp);

        mToolbarLayout = (TextInputLayout) findViewById(R.id.toolbarLayout);
        Assert.assertTrue(mToolbarLayout != null);

        mToolbarEditText = (EditText) findViewById(R.id.toolbarEditText);
        Assert.assertTrue(mToolbarEditText != null);

        FloatingActionButton showProjectFab = (FloatingActionButton) findViewById(R.id.show_project_fab);
        Assert.assertTrue(showProjectFab != null);

        mToolbarEditText.addTextChangedListener(new TextWatcher() {
            private boolean mSkip = true;

            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {

            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {

            }

            @Override
            public void afterTextChanged(Editable s) {
                if (mSkip) {
                    mSkip = false;
                    return;
                }

                updateError();
            }
        });

        Intent intent = getIntent();
        if (intent.hasExtra(PROJECT_ID_KEY)) {
            mProjectId = intent.getStringExtra(PROJECT_ID_KEY);
            Assert.assertTrue(!TextUtils.isEmpty(mProjectId));
        }

        mUserListFragment = (UserListFragment) getSupportFragmentManager().findFragmentById(R.id.show_project_frame);
        if (mUserListFragment == null) {
            mUserListFragment = UserListFragment.newInstance();

            getSupportFragmentManager()
                    .beginTransaction()
                    .add(R.id.show_project_frame, mUserListFragment)
                    .commit();
        }
        mUserListFragment.setFab(showProjectFab);

        DiscardDialogFragment discardDialogFragment = (DiscardDialogFragment) getSupportFragmentManager().findFragmentByTag(DISCARD_TAG);
        if (discardDialogFragment != null)
            discardDialogFragment.setDiscardDialogListener(mDiscardDialogListener);

        getSupportLoaderManager().initLoader(0, null, this);
    }

    @Override
    public Loader<ShowProjectLoader.Data> onCreateLoader(int id, Bundle args) {
        return new ShowProjectLoader(this, mProjectId);
    }

    @Override
    public void onLoadFinished(Loader<ShowProjectLoader.Data> loader, ShowProjectLoader.Data data) {
        Assert.assertTrue(data != null);

        mData = data;

        if (mSavedInstanceState == null) {
            mToolbarEditText.setText(data.mName);
        } else {
            mSavedInstanceState = null;
        }

        mToolbarLayout.setVisibility(View.VISIBLE);
        mToolbarLayout.setHintAnimationEnabled(true);

        invalidateOptionsMenu();

        mUserListFragment.initialize(mProjectId, data);
    }

    @Override
    public void onLoaderReset(Loader<ShowProjectLoader.Data> loader) {

    }

    @Override
    public void onBackPressed() {
        if (tryClose())
            super.onBackPressed();
    }

    private boolean tryClose() {
        if (dataChanged()) {
            DiscardDialogFragment discardDialogFragment = DiscardDialogFragment.newInstance();
            discardDialogFragment.setDiscardDialogListener(mDiscardDialogListener);
            discardDialogFragment.show(getSupportFragmentManager(), DISCARD_TAG);

            return false;
        } else {
            return true;
        }
    }

    @SuppressWarnings("SimplifiableIfStatement")
    private boolean dataChanged() {
        if (mData == null)
            return false;

        if (TextUtils.isEmpty(mToolbarEditText.getText()) != TextUtils.isEmpty(mData.mName))
            return true;

        if (!TextUtils.isEmpty(mToolbarEditText.getText()) && !mToolbarEditText.getText().toString().equals(mData.mName))
            return true;

        return mUserListFragment.dataChanged();
    }

    private boolean updateError() {
        Assert.assertTrue(mData != null);
        Assert.assertTrue(mToolbarEditText != null);
        Assert.assertTrue(mToolbarLayout != null);

        if (TextUtils.isEmpty(mToolbarEditText.getText())) {
            mToolbarLayout.setError(getString(R.string.nameError));

            return true;
        } else {
            mToolbarLayout.setError(null);

            return false;
        }
    }
}

File: app/src/main/java/com/krystianwsul/checkme/gui/projects/ShowProjectActivity.kt
package com.krystianwsul.checkme.gui.projects

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.os.Parcelable
import android.text.Editable
import android.text.TextWatcher
import android.view.Menu
import android.view.MenuItem
import android.view.View
import com.krystianwsul.checkme.R
import com.krystianwsul.checkme.databinding.ActivityShowProjectBinding
import com.krystianwsul.checkme.databinding.BottomBinding
import com.krystianwsul.checkme.gui.base.AbstractActivity
import com.krystianwsul.checkme.gui.dialogs.ConfirmDialogFragment
import com.krystianwsul.checkme.gui.friends.UserListFragment
import com.krystianwsul.checkme.gui.utils.BottomFabMenuDelegate
import com.krystianwsul.checkme.utils.tryGetFragment
import com.krystianwsul.checkme.viewmodels.ShowProjectViewModel
import com.krystianwsul.checkme.viewmodels.getViewModel
import com.krystianwsul.common.utils.ProjectKey
import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers
import io.reactivex.rxjava3.kotlin.addTo
import io.reactivex.rxjava3.kotlin.plusAssign

class ShowProjectActivity : AbstractActivity(), UserListFragment.UserListListener {

    companion object {

        private const val PROJECT_ID_KEY = "projectId"

        private const val DISCARD_TAG = "discard"

        private const val KEY_BOTTOM_FAB_MENU_DELEGATE_STATE = "bottomFabMenuDelegateState"

        fun newIntent(
            context: Context,
            projectId: ProjectKey.Shared,
        ) = Intent(context, ShowProjectActivity::class.java).apply {
            putExtra(PROJECT_ID_KEY, projectId as Parcelable)
        }

        fun newIntent(context: Context) = Intent(context, ShowProjectActivity::class.java)
    }

    private var projectId: ProjectKey.Shared? = null

    private var data: ShowProjectViewModel.Data? = null

    private lateinit var userListFragment: UserListFragment

    private var savedInstanceState: Bundle? = null

    private val discardDialogListener: (Parcelable?) -> Unit = { finish() }

    private lateinit var showProjectViewModel: ShowProjectViewModel

    private var selectAllVisible = false

    private lateinit var binding: ActivityShowProjectBinding
    private lateinit var bottomBinding: BottomBinding

    private lateinit var bottomFabMenuDelegate: BottomFabMenuDelegate

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_save, menu)
        return true
    }

    override fun onPrepareOptionsMenu(menu: Menu): Boolean {
        menu.findItem(R.id.action_save).isVisible = data != null
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.action_save -> {
                check(data != null)

                if (updateError())
                    return true

                showProjectViewModel.stop()

                userListFragment.save(binding.showProjectToolbarEditTextInclude.toolbarEditText.text.toString())
                        .observeOn(AndroidSchedulers.mainThread())
                        .subscribe { finish() }
                        .addTo(createDisposable)
            }
            android.R.id.home -> {
                if (tryClose())
                    finish()
            }
            else -> throw UnsupportedOperationException()
        }
        return true
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        binding = ActivityShowProjectBinding.inflate(layoutInflater)
        bottomBinding = BottomBinding.bind(binding.root)
        setContentView(binding.root)

        bottomFabMenuDelegate = BottomFabMenuDelegate(
            bottomBinding,
            binding.showProjectCoordinator,
            this,
            savedInstanceState?.getParcelable(KEY_BOTTOM_FAB_MENU_DELEGATE_STATE),
        )

        this.savedInstanceState = savedInstanceState

        setSupportActionBar(binding.showProjectToolbarEditTextInclude.toolbar)

        supportActionBar!!.run {
            setDisplayHomeAsUpEnabled(true)
            setHomeAsUpIndicator(R.drawable.ic_close_white_24dp)
        }

        initBottomBar()

        binding.showProjectToolbarEditTextInclude
                .toolbarEditText
                .addTextChangedListener(object : TextWatcher {

                    private var skip = true

                    override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) = Unit

                    override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) = Unit

                    override fun afterTextChanged(s: Editable) {
                        if (skip) skip = false else updateError()
                    }
                })

        if (intent.hasExtra(PROJECT_ID_KEY)) projectId = intent.getParcelableExtra(PROJECT_ID_KEY)

        userListFragment = supportFragmentManager.findFragmentById(R.id.show_project_frame) as? UserListFragment
                ?: UserListFragment.newInstance().also {
                    supportFragmentManager
                            .beginTransaction()
                            .add(R.id.show_project_frame, it)
                            .commit()
                }

        userListFragment.setFab(bottomFabMenuDelegate.fabDelegate)

        tryGetFragment<ConfirmDialogFragment>(DISCARD_TAG)?.listener = discardDialogListener

        showProjectViewModel = getViewModel<ShowProjectViewModel>().apply {
            start(projectId)

            createDisposable += data.subscribe { onLoadFinished(it) }
        }
    }

    private fun onLoadFinished(data: ShowProjectViewModel.Data) {
        this.data = data

        if (savedInstanceState == null) {
            binding.showProjectToolbarEditTextInclude
                    .toolbarEditText
                    .setText(data.name)
        } else {
            savedInstanceState = null
        }

        binding.showProjectToolbarEditTextInclude
                .toolbarLayout
                .apply {
                    visibility = View.VISIBLE
                    isHintAnimationEnabled = true
                }

        invalidateOptionsMenu()

        userListFragment.initialize(projectId, data)
    }

    override fun onBackPressed() {
        if (tryClose())
            super.onBackPressed()
    }

    private fun tryClose(): Boolean {
        return if (dataChanged()) {
            ConfirmDialogFragment.newInstance(ConfirmDialogFragment.Parameters.Discard).also {
                it.listener = discardDialogListener
                it.show(supportFragmentManager, DISCARD_TAG)
            }

            false
        } else {
            true
        }
    }

    private fun dataChanged(): Boolean {
        if (data == null)
            return false

        if (binding.showProjectToolbarEditTextInclude.toolbarEditText.text.isNullOrEmpty() != data!!.name.isNullOrEmpty())
            return true

        return if (!binding.showProjectToolbarEditTextInclude
                        .toolbarEditText
                        .text
                        .isNullOrEmpty()
                && binding.showProjectToolbarEditTextInclude
                        .toolbarEditText
                        .text
                        .toString() != data!!.name
        ) {
            true
        } else {
            userListFragment.dataChanged()
        }
    }

    private fun updateError(): Boolean {
        check(data != null)

        return if (binding.showProjectToolbarEditTextInclude.toolbarEditText.text.isNullOrEmpty()) {
            binding.showProjectToolbarEditTextInclude
                    .toolbarLayout
                    .error = getString(R.string.nameError)

            true
        } else {
            binding.showProjectToolbarEditTextInclude
                    .toolbarLayout
                    .error = null

            false
        }
    }

    override fun initBottomBar() {
        bottomBinding.bottomAppBar.apply {
            replaceMenu(R.menu.menu_select_all)

            setOnMenuItemClickListener { item ->
                check(item.itemId == R.id.action_select_all)

                userListFragment.treeViewAdapter.selectAll()

                true
            }
        }
    }

    override fun getBottomBar() = bottomBinding.bottomAppBar

    private fun updateBottomMenu() {
        bottomBinding.bottomAppBar
                .menu
            .findItem(R.id.action_select_all)
            ?.isVisible = selectAllVisible
    }

    override fun setUserSelectAllVisibility(selectAllVisible: Boolean) {
        this.selectAllVisible = selectAllVisible

        updateBottomMenu()
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)

        outState.putParcelable(KEY_BOTTOM_FAB_MENU_DELEGATE_STATE, bottomFabMenuDelegate.state)
    }

    override val snackbarParent get() = binding.showProjectCoordinator
}
